x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -168,47 +229,58 @@ public void applyClusterState(ClusterChangedEvent event) {
             this.autoFollowPatternsCountDown = new CountDown(autoFollowMetadata.getPatterns().size());
+            this.clusterAliasResults = new AtomicArray<>(autoFollowMetadata.getPatterns().size());
         }
 
         void autoFollowIndices() {
+            int i = 0;
             for (Map.Entry<String, AutoFollowPattern> entry : autoFollowMetadata.getPatterns().entrySet()) {
-                String clusterAlias = entry.getKey();
-                AutoFollowPattern autoFollowPattern = entry.getValue();
-                List<String> followedIndices = autoFollowMetadata.getFollowedLeaderIndexUUIDs().get(clusterAlias);
+                final int slot = i;
+                final String clusterAlias = entry.getKey();
+                final AutoFollowPattern autoFollowPattern = entry.getValue();
+                final List<String> followedIndices = autoFollowMetadata.getFollowedLeaderIndexUUIDs().get(clusterAlias);
 
                 getLeaderClusterState(clusterAlias, (leaderClusterState, e) -> {
                     if (leaderClusterState != null) {
                         assert e == null;
-                        handleClusterAlias(clusterAlias, autoFollowPattern, followedIndices, leaderClusterState);
+                        handleClusterAlias(slot, clusterAlias, autoFollowPattern, followedIndices, leaderClusterState);
                     } else {
-                        finalise(e);
+                        finalise(slot, new AutoFollowResult(clusterAlias, e));
                     }
                 });
+                i++;
             }
         }
 
-        private void handleClusterAlias(String clusterAlias, AutoFollowPattern autoFollowPattern,
-                                List<String> followedIndexUUIDs, ClusterState leaderClusterState) {
+        private void handleClusterAlias(
+                int clusterAliasSlot,
+                String clusterAlias,
+                AutoFollowPattern autoFollowPattern,
+                List<String> followedIndexUUIDs,
+                ClusterState leaderClusterState
+        ) {
             final List<Index> leaderIndicesToFollow =
                 getLeaderIndicesToFollow(autoFollowPattern, leaderClusterState, followerClusterState, followedIndexUUIDs);
             if (leaderIndicesToFollow.isEmpty()) {
-                finalise(null);
+                finalise(clusterAliasSlot, new AutoFollowResult(clusterAlias));
             } else {
                 final CountDown leaderIndicesCountDown = new CountDown(leaderIndicesToFollow.size());
-                final AtomicReference<Exception> leaderIndicesErrorHolder = new AtomicReference<>();
-                for (Index indexToFollow : leaderIndicesToFollow) {
+                final AtomicArray<Tuple<Index, Exception>> results = new AtomicArray<>(leaderIndicesToFollow.size());
+                for (int i = 0; i < leaderIndicesToFollow.size(); i++) {
+                    final int slot = i;
```

Comment:
I am wondering if we can use another data-structure (instead of AtomicArray) to avoid passing the slot-index around. It may be less error-prone because we now have two slot-indexes (clusterAliasSlot and slot) in `handleClusterAlias` method.

Reply:
maybe it would be better if we just do not pass the `clusterAliasSlot` around. Let me see if this would be cleaner.

Reply:
I think is better now: https://github.com/elastic/elasticsearch/pull/33684/commits/90fb198b20468bba0f1dbcfd5b8827c48b14fba0 ?

Reply:
@dnhatn The splitup of `handleClusterAlias()` method like we discussed privately: f501c10c698bc5eb77ddcdaef8640531e268d6f7